commit 966115c9461c1030f5a311924ead0ea2207634b4 (refs/changes/75/2775/8)
Author: Tim Kordas <tim.kordas@joyent.com>
Date:   2017-10-17T16:47:00-07:00 (2 years ago)
    
    joyent/node-cueball#132 Expose counters and simple metrics from CueBallConnectionPool
    Reviewed by: Cody Peter Mello <cody.mello@joyent.com>

diff --git a/CHANGES.adoc b/CHANGES.adoc
index f7394f2..a2ef7fb 100644
--- a/CHANGES.adoc
+++ b/CHANGES.adoc
@@ -6,6 +6,17 @@ toc::[]
 
 ## v2.x
 
+### v2.5.0
+
+New minor release, due to addition of new API.
+
+API Changes:
+
+ - The `Pool` class now has a method `getStats()`, which returns a snapshot
+   of the Pool instance's counters along with four additional numbers:
+   total connections, idle connections, partly-open connections, and the
+   number of requests queued on available connections (#132)
+
 ### v2.4.0
 
 New minor release, due to addition of new API.
diff --git a/lib/pool.js b/lib/pool.js
index 84cab63..e72d0b9 100644
--- a/lib/pool.js
+++ b/lib/pool.js
@@ -729,6 +729,26 @@ CueBallConnectionPool.prototype.printConnections = function () {
 	console.log('dead:', self.p_dead);
 };
 
+CueBallConnectionPool.prototype.getStats = function () {
+	var self = this;
+
+	// Get a snapshot of current counter values.
+	var counters = {};
+	Object.keys(this.p_counters).forEach(function (k) {
+		counters[k] = self.p_counters[k];
+	});
+
+	var stats = {
+		'counters': counters,
+		'totalConnections': Object.keys(self.p_connections).length,
+		'idleConnections': self.p_idleq.length,
+		'pendingConnections': self.p_initq.length,
+		'waiterCount': self.p_waiters.length
+	};
+
+	return (stats);
+};
+
 CueBallConnectionPool.prototype.claim = function (options, cb) {
 	var self = this;
 	var done = false;
diff --git a/package.json b/package.json
index e1bd03f..d6cdaf4 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "cueball",
-  "version": "2.4.0",
+  "version": "2.5.0",
   "description": "manage a pool of connections to a multi-node service where nodes are listed in DNS",
   "main": "lib/index.js",
   "dependencies": {
